packages/calcite-components/src/components/carousel/carousel.tsx

```diff
@@ -0,0 +1,353 @@
+    connectMessages(this);
+  }
+
+  componentDidLoad(): void {
+    setComponentLoaded(this);
+  }
+
+  componentDidRender(): void {
+    updateHostInteraction(this);
+  }
+
+  disconnectedCallback(): void {
+    disconnectInteractive(this);
+    disconnectLocalized(this);
+    disconnectMessages(this);
+  }
+
+  async componentWillLoad(): Promise<void> {
+    setUpLoadableComponent(this);
+    await setUpMessages(this);
+  }
+
+  // --------------------------------------------------------------------------
+  //
+  //  Private Properties
+  //
+  // --------------------------------------------------------------------------
+
+  @Element() el: HTMLCalciteCarouselElement;
+
+  @State() activeIndex: number;
+
+  @State() items: HTMLCalciteCarouselItemElement[];
+
+  @State() direction: "advancing" | "retreating";
+
+  @State() defaultMessages: CarouselMessages;
+
+  @State() effectiveLocale = "";
+
+  @Watch("effectiveLocale")
+  async effectiveLocaleChange(): Promise<void> {
+    await updateMessages(this, this.effectiveLocale);
+  }
+
+  private container: HTMLDivElement;
+
+  // --------------------------------------------------------------------------
+  //
+  //  Events
+  //
+  // --------------------------------------------------------------------------
+
+  /** Fires when the selected carousel item changes. */
+  @Event({ cancelable: false }) calciteCarouselChange: EventEmitter<void>;
+
+  // --------------------------------------------------------------------------
+  //
+  //  Private Methods
+  //
+  // --------------------------------------------------------------------------
+
+  private updateItems = (event: Event): void => {
+    const items = slotChangeGetAssignedElements(event) as HTMLCalciteCarouselItemElement[];
+
+    if (items.length < 1) {
+      return;
+    }
+
+    const activeItemIndex = items?.findIndex((item) => item.active);
+    const requestedActiveIndex = activeItemIndex > -1 ? activeItemIndex : 0;
+
+    this.items = items;
+    this.setSelectedItem(false, requestedActiveIndex);
+  };
+
+  private setSelectedItem = (emit: boolean, requestedIndex: number): void => {
+    this.items?.forEach((el, index) => {
+      const isMatch = requestedIndex === index;
+      el.active = isMatch;
+      if (isMatch) {
+        this.selectedItem = el;
+        this.activeIndex = index;
+      }
+    });
+
+    if (emit) {
+      this.calciteCarouselChange.emit();
+    }
+  };
+
+  private nextItem = (): void => {
+    this.direction = "advancing";
+    const nextIndex = this.activeIndex === this.items?.length - 1 ? 0 : this.activeIndex + 1;
+    this.setSelectedItem(true, nextIndex);
+  };
+
+  private previousItem = (): void => {
+    this.direction = "retreating";
+    const previousIndex = this.activeIndex === 0 ? this.items?.length - 1 : this.activeIndex - 1;
+    this.setSelectedItem(true, previousIndex);
+  };
+
+  private keyDownHandler = (event: KeyboardEvent): void => {
+    if (event.target !== this.container) {
```

Comment:
```suggestion
    if (event.target !== this.container || event.defaultPrevented) {
```

Reply:
Shouldn't keyboard nav work on any element within the component so you don't have to be on the container? Seems like we should support bubbling events here to allow for navigation. Then if a component doesn't want the functionality they preventdefault or stop bubbling.

Reply:
good question for @jcfranco. We should establish a pattern here. 

I know for things like modal, you can be anywhere in the component and esc will close it. I would expect this to work the same way.

Reply:
> good question for @jcfranco. We should establish a pattern here.

Sorry, I missed this. What was the question? 😅